src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalPane.java

```diff
@@ -351,22 +438,56 @@ public void onSwitchToTerminal(SwitchToTerminalEvent event)
       TerminalMetadata existingTerminal = terminals_.getMetadataForHandle(handle);
       if (existingTerminal != null)
       {
-         startTerminal(existingTerminal.getSequence(), handle);
+         startTerminal(existingTerminal.getSequence(),
+                       handle,
+                       existingTerminal.getCaption(),
+                       existingTerminal.getTitle());
          return;
       }
 
       Debug.logWarning("Tried to switch to unknown terminal handle");
    }
 
    @Override
-   public void onTerminalCaption(TerminalCaptionEvent event)
+   public void onTerminalTitle(TerminalTitleEvent event)
    {
+      // Notification from a TerminalSession that it changed its title
       TerminalSession visibleTerm = getSelectedTerminal();
-      TerminalSession captionTerm = event.getTerminalSession();
+      TerminalSession retitledTerm = event.getTerminalSession();
       if (visibleTerm != null && visibleTerm.getHandle().equals(
-            captionTerm.getHandle()))
+            retitledTerm.getHandle()))
+      {
+         // update the toolbar label if currently displayed terminal has changed
+         // its title
+         setTerminalTitle(retitledTerm.getTitle());
+      }
+
+      TerminalMetadata currentData = terminals_.getMetadataForHandle(retitledTerm.getHandle());
+      if (currentData == null)
+         return;
+
+      if (!currentData.getTitle().equals(retitledTerm.getTitle()))
       {
-         setTerminalCaption(captionTerm.getCaption());
+         // Update local metadata
+         terminals_.addTerminal(
+               new TerminalMetadata(
+                     currentData.getHandle(),
+                     currentData.getCaption(),
+                     retitledTerm.getTitle(),
+                     currentData.getSequence()));
+
+         // update server
+         server_.processSetTitle(currentData.getHandle(), 
+               retitledTerm.getTitle(),
+               new VoidServerRequestCallback()
+               {
+                  @Override
+                  public void onError(ServerError error)
+                  {
+                     // failed; this might mean we show the wrong title after
+                     // a reset, but it will update itself fairly quickly
```

Comment:
If this error isn't expected during the normal course of operations, it should be logged to the console (see `Debug.logError`)